driver-embedded/src/main/com/mongodb/embedded/client/MongoDBCAPIHelper.java

```diff
@@ -64,108 +67,103 @@ static synchronized void init(final MongoEmbeddedSettings mongoEmbeddedSettings)
         }
     }
 
-    static void db_pump(final Pointer db) {
+    static void instance_destroy(final Pointer instance) {
         checkInitialized();
         try {
-            validateErrorCode(mongoDBCAPI.libmongodbcapi_db_pump(db));
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_instance_destroy(instance, status));
         } catch (Throwable t) {
-            throw createError("db_pump", t);
+            throw createError("instance_destroy", t);
         }
     }
 
-    static Pointer db_client_new(final Pointer db) {
+    static Pointer create_client(final Pointer instance) {
         checkInitialized();
         try {
-            return mongoDBCAPI.libmongodbcapi_db_client_new(db);
+            return validatePointerCreated(mongoDBCAPI.libmongodbcapi_client_create(instance, status));
         } catch (Throwable t) {
-            throw createError("db_client_new", t);
+            throw createError("client_create", t);
         }
     }
 
-    static void db_client_destroy(final Pointer client) {
+    static void client_destroy(final Pointer client) {
         checkInitialized();
         try {
-            mongoDBCAPI.libmongodbcapi_db_client_destroy(client);
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_client_destroy(client, status));
         } catch (Throwable t) {
-            throw createError("db_client_destroy", t);
+            throw createError("client_destroy", t);
         }
     }
 
-    static void db_client_wire_protocol_rpc(final Pointer client, final byte[] input, final int inputSize,
-                                                         final PointerByReference output, final IntByReference outputSize) {
+    static void client_invoke(final Pointer client, final byte[] input, final PointerByReference output, final IntByReference outputSize) {
         checkInitialized();
         try {
-            validateErrorCode(mongoDBCAPI.libmongodbcapi_db_client_wire_protocol_rpc(client, input, inputSize, output, outputSize));
+            validateErrorCode(mongoDBCAPI.libmongodbcapi_client_invoke(client, input, input.length, output, outputSize, status));
         } catch (Throwable t) {
-            throw createError("db_client_wire_protocol_rpc", t);
+            throw createError("client_invoke", t);
         }
     }
 
-    static int get_last_error() {
-        checkInitialized();
-        try {
-            return mongoDBCAPI.libmongodbcapi_get_last_error();
-        } catch (Throwable t) {
-            throw createError("get_last_error", t);
-        }
+    private static MongoClientException createError(final String methodName, final Throwable t) {
+        return new MongoClientException(format("Error from embedded server when calling '%s': %s", methodName, t.getMessage()), t);
+    }
+
+    private static MongoClientException createErrorFromStatus() {
+        throw new MongoClientException(mongoDBCAPI.libmongodbcapi_status_get_explanation(status));
```

Comment:
The exception should include all three of:

* libmongodbcapi_status_get_error (might be the same as what's returned from the erroring method)
* libmongodbcapi_status_get_explanation
* libmongodbcapi_status_get_code (not really sure what this is, but since server wants to report it we should include it).

I think also we should consider a new MongoException subclass that has these three things as explicit properties, as well as a Throwable.message that concats them all together, similar to MongoCommandException.

Reply:
Done.